ddprof-lib/src/main/cpp/wallClock_jvmti.cpp

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2018 Andrei Pangin
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <math.h>
+#include <random>
+#include "debugSupport.h"
+#include "wallClock_jvmti.h"
+#include "profiler.h"
+#include "stackFrame.h"
+#include "context.h"
+#include "log.h"
+#include "thread.h"
+#include "tsc.h"
+#include "vmStructs.h"
+#include "reservoir_sampler.h"
+#include <jvmti.h>
+
+std::atomic<bool> JvmtiWallClockSampler::_enabled{false};
+
+Error JvmtiWallClockSampler::start(Arguments &args) {
+    int interval = args._event != NULL ? args._interval : args._wall;
+    if (interval < 0) {
+        return Error("interval must be positive");
+    }
+    _interval = interval ? interval : DEFAULT_WALL_INTERVAL;
+
+    _reservoir_size =
+            args._wall_threads_per_tick ?
+            args._wall_threads_per_tick :
+            DEFAULT_WALL_THREADS_PER_TICK;
+
+    _running = true;
+
+    if (pthread_create(&_thread, NULL, threadEntry, this) != 0) {
+        return Error("Unable to create timer thread");
+    }
+
+    return Error::OK;
+}
+
+void JvmtiWallClockSampler::stop() {
+    _running.store(false);
+    // the thread join ensures we wait for the thread to finish before returning (and possibly removing the object)
+    pthread_kill(_thread, WAKEUP_SIGNAL);
+    int res = pthread_join(_thread, NULL);
+    if(res != 0) {
+        Log::warn("Unable to join WallClock thread on stop %d", res);
+    }
+}
+
+void JvmtiWallClockSampler::timerLoop() {
+    ThreadFilter* thread_filter = Profiler::instance()->threadFilter();
+    if (!_enabled.load(std::memory_order_acquire) || !thread_filter->enabled()) {
+        return;
+    }
+    JNIEnv* jni = VM::attachThread("datadog-wallclock-sampler");
+    jvmtiEnv* jvmti = VM::jvmti();
+    const int max_frames = Profiler::instance()->max_stack_depth();
+    jvmtiFrameInfo* frames = frames = new jvmtiFrameInfo[max_frames];
+    std::vector<int> tids;
+    tids.reserve(_reservoir_size);
+    std::vector<jthread> javaThreads;
+    javaThreads.reserve(_reservoir_size);
+    ReservoirSampler<int> reservoir(_reservoir_size);
+    int self = OS::threadId();
+    u64 counter = 0;
+
+
+    while (_running.load(std::memory_order_relaxed)) {
+        int numFailures = 0;
+        int threadsAlreadyExited = 0;
+        std::vector<int> sample = reservoir.sample(tids);
```

Comment:
Where is the `tids` vector filled with the TIDs?
I might be blind but I can't seem to find it :(

Reply:
In the ASGCT version I see the thread_filter taking care of that - https://github.com/DataDog/java-profiler/pull/118/files#diff-a368edb3b62aef3c92f4e699fc69e1a6fabed5d158b808dabda9af2018073ddaR158-R170

Reply:
good catch, the test wasn't testing what I thought it was. I've added some extra tests to verify that the profiler under test is the one in use.